### PR TITLE
feat: add a custom storage options param

### LINKED ISSUE #236: Inject custom s3 connection configs (credentials)
## 🚀 Feature

- Inject custom s3 connection config into the StreamingDataset

### Motivation

Thanks you for the awesome project! I primarily use litdata in a local environment, connecting to S3 to fetch data. However, managing multiple S3 connections for services like mlflow has become quite challenging, especially when handling AWS S3 credentials through environment variables.

To address this, **it would be great if we could inject custom S3 connection configurations into the streaming dataset.** Currently, it appears that the S3Client is running globally.

### PR DESCRIPTION
**Pull request by csy1204**
<details>
  <summary><b>Before submitting</b></summary>

- [ ] Was this discussed/agreed via a Github issue? (no need for typos and docs improvements)
- [ ] Did you read the [contributor guideline](https://github.com/Lightning-AI/lit-data/blob/main/.github/CONTRIBUTING.md), Pull Request section?
- [ ] Did you make sure to update the docs?
- [ ] Did you write any new necessary tests?

</details>

## What does this PR do?

Fixes #236 (issue).

``` py
import litdata as ld

storage_options = {
    "endpoint_url": url,
    "aws_access_key_id": id,
    "aws_secret_access_key": pw,
}

s3_path = "s3://bucket/test_dataset"
dataset = ld.StreamingDataset(s3_path, storage_options=storage_options)
```

## PR review

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in GitHub issues there's a high chance it will not be merged.

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: tests/streaming/test_client.py [hunk 1]
```diff
@@ -6,6 +6,34 @@
 from litdata.streaming import client
 
 
+def test_s3_client_with_storage_options(monkeypatch):
```
Comment (tchaton) [inline tests/streaming/test_client.py:9]: Can you add a test for GCP ?

File: src/litdata/streaming/client.py [hunk 2]
```diff
@@ -38,7 +39,9 @@ def _create_client(self) -> None:
 
         if has_shared_credentials_file or not _IS_IN_STUDIO:
             self._client = boto3.client(
-                "s3", config=botocore.config.Config(retries={"max_attempts": 1000, "mode": "adaptive"})
+                "s3",
+                **self._storage_options,
```
Comment (tchaton) [inline src/litdata/streaming/client.py:43]: Maybe, we should put this one after the config object, so users can override it with their own. wdyt ?

**Review by tchaton — COMMENTED** [review]
Hey @csy1204 Looks great. Could you add an example in the README ?

File: src/litdata/streaming/client.py (hunk 2 above)
Comment (csy1204) [reply src/litdata/streaming/client.py]: that's reasonable, so I fixed it! https://github.com/Lightning-AI/litdata/pull/246/commits/be80ef6d4d7fa774a7c280484fc1e80524ebd32b

File: README.md [hunk 3]
```diff
@@ -209,6 +209,21 @@ for batch in dataloader:
 
 ```
 
+
+Additionally, You can inject client connection settings for [S3](https://boto3.amazonaws.com/v1/documentation/api/latest/reference/core/session.html#boto3.session.Session.client) or GCP when initializing your dataset. This is useful for specifying custom endpoints and credentials.
```
Comment (tchaton) [inline README.md:213]:
```suggestion
Additionally, you can inject client connection settings for [S3](https://boto3.amazonaws.com/v1/documentation/api/latest/reference/core/session.html#boto3.session.Session.client) or GCP when initializing your dataset. This is useful for specifying custom endpoints and credentials per dataset.
```

**Review by tchaton — APPROVED** [review]

**Comment by csy1204** [comment]
@tchaton Thanks for your reviews.
I've added the example of this right after `✅ Stream large cloud datasets` documentation. Feel free to edit my example 😃

**Comment by tchaton** [comment]
Thanks @csy1204 for your contribution ;) Feel free to continue.

**Pull request merged**